### PR TITLE
use HTTPS only

### LINKED ISSUE #12: use HTTPS only

### PR DESCRIPTION
**Pull request by afeld**
Closes #12.

This pull request adds an nginx proxy in front of Jenkins to serve requests - this seemed more trustworthy and configurable than relying on Jenkins to do HTTPS termination itself. While the example security group setup leaves port 80 open, this is simply to redirect to 443. My test deployment, using a self-signed cert (hence why it says "Not Secure"):

dashboard__jenkins__and_jdauphant_ansible-role-ssl-certs__generate_and_or_deploy_ssl_certificate

Note that all the files under `test/roles/` are vendored third-party dependencies.

### CODE REVIEW

File: tasks/nginx.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  uri:
+    url: https://localhost
+    follow_redirects: none
+    return_content: yes
+    # TODO have it use the test cert
+    validate_certs: no
+  register: https_req
+- name: Fail if HTTPS response isn't from Jenkins
+  fail:
+  when: "'Jenkins' not in https_req.content"
```
Comment (afeld) [inline tasks/nginx.yml:41]: These are examples of building [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_(software)) into roles themselves, an approach I have been very happy with.

File: templates/vhost.conf [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    proxy_set_header        X-Real-IP $remote_addr;
+    proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_set_header        X-Forwarded-Proto $scheme;
+    proxy_redirect http:// https://;
+    proxy_pass              http://jenkins;
+    # Required for new HTTP-based CLI
+    proxy_http_version 1.1;
+    proxy_request_buffering off;
+  }
+}
```
Comment (afeld) [inline templates/vhost.conf:28]: This is the important part. This file gets `include`d into an `nginx.conf`, which is generated by the `geerlingguy.nginx` role.

Comment (afeld) [reply templates/vhost.conf:28]: I pretty much took contents of this file from the link above, so would love some more in-depth review and suggestions here. In particular: should each of the `server` blocks have `server_name` directives, or does it matter?

**Review by yfuksenko — APPROVED** [review]

File: templates/vhost.conf (hunk 2 above)
Comment (afeld) [reply templates/vhost.conf:28]: Per @maverickquant:

* [x] Specify `server_name`

Comment (afeld) [reply templates/vhost.conf:28]: Put other nginx hardening feedback/questions in https://github.com/GSA/jenkins-deploy/issues/22.

**Comment by afeld** [comment]
@yfuksenko This pull request has changed a bit - mind taking another look? I suggest we take care of any additional hardening as follow-up(s).

File: meta/main.yml [hunk 3]
```diff
@@ -43,15 +43,36 @@ galaxy_info:
...
+  - role: geerlingguy.nginx
+    nginx_remove_default_vhost: true
+    nginx_server_tokens: "off"
+    # use EPEL version
+    nginx_yum_repo_enabled: false
+    # https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+behind+an+NGinX+reverse+proxy#JenkinsbehindanNGinXreverseproxy-RunningfromasubdomainwithSSL
+    nginx_upstreams:
+      - name: jenkins
+        servers:
+          - "{{ jenkins_hostname }}:{{ jenkins_http_port }} fail_timeout=0"
```
Comment (afeld) [inline meta/main.yml:71]: Move to the `vhost.conf` to make this all easier to follow.

Comment (afeld) [reply meta/main.yml:71]: Hmm, seems we need to use upstreams to get the various [configuration options](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#directives), which we are using here, and seem like others could add value. Going to leave it here for now, though we can certainly revisit later.

**Comment by afeld** [comment]
Got verbal re-approval from @yfuksenko, so merging!

**Pull request merged**